sdk/src/account_data.rs

```diff
@@ -0,0 +1,317 @@
+use serde::de::{Deserialize, Deserializer};
+use serde::ser::{Serialize, Serializer};
+use solana_program::sysvar::{Sysvar, SysvarEnum};
+use std::{
+    any::TypeId,
+    collections::{hash_map::Entry, HashMap},
+    convert::TryFrom,
+    iter::FromIterator,
+    ops::{Deref, DerefMut, Index, IndexMut},
+    slice::SliceIndex,
+    sync::RwLock,
+};
+
+#[derive(Debug, Default, AbiExample)]
+pub struct AccountData {
+    data: Vec<u8>,
+    cache: RwLock<HashMap<TypeId, Option<SysvarEnum>>>,
```

Comment:
You are right that theoretically It should be only one sysvar type.  
But from just a bytes array `Vec<u8>` we cannot say what the `Sysvar` type is. That is `bincode::deserialize<S>(data)` may happily de-serialize for two different sysvar types `S1` and `S2`, because bincode does not encode any schema. Obviously one of those two invocations is logical error but we cannot distinguish them from the bytes array.

This is exactly the same behavior as the current `from_account` implementation:
https://github.com/solana-labs/solana/blob/61c7ce857/sdk/src/account.rs#L122-L125
just storing the result in the hashmap.